### DISCUSSION TITLE
[Question] Directly set state can trigger rerender?

### DISCUSSION THREAD

**Discussion by ShinyLeee** [pmndrs/zustand#859] [Q&A] [answered]
I found we can directly set state in the official demo without return a new object, update nested object is the same. But in [official recipes](https://docs.pmnd.rs/zustand/recipes) i did't found any thing like this other than immer. So my question is can we do this? 

`const useStore = create((set) => ({
  count: 1,
  inc: () => set((state) => { state.count += 1 }),
}))`

https://codesandbox.io/s/hopeful-wiles-d58fzs?file=/src/App.js

**Comment by dai-shi** [comment, marked as answer]
No, you can't. It should be updated immutably (unless using immer). Mutation may cause unexpected behavior.

**Reply by ShinyLeee** [reply]
Well, I think so. 

This thing start from when i add immer middleware and use Redux Devtool time travel some error happened, when i remove immer middleware found this can rerender as well...

```
Unhandled Runtime Error
Error: [Immer] The first or second argument to `produce` must be a function
```
